### PR TITLE
refactor api models; basic support for gpt3.5-turbo

### PR DESCRIPTION
**Pull request by tannonk**
This PR includes:

1. refactor the API model inference script – we provide a wrapper class `API_LLM` that replicates the behaviour of our `LLM` class for Hugging Face models. 
2. experiment configs for OpenAI models for easy replication inference runs
3. bash/sbatch wrappers for running inference and evaluation pipeline easily

Note, the integration of Chat models is v basic!

@dennlinger, any feedback on this would be great!

### CODE REVIEW

File: inference_API_models.py
Comment (tannonk) [inline inference_API_models.py:1]: functionality has been integrated into `inference.py` and `llm_inference.py`

File: slurm_scripts/run_inference_all.sh
Comment (tannonk) [inline slurm_scripts/run_inference_all.sh:1]: redundant

File: llm_inference.py
Comment (tannonk) [inline llm_inference.py:1]: Would be nice to log the number of tokens processed in order to keep track of costs too.

File: exp_configs/cluster/openai-gpt-3.5-turbo.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+    "use_slurm": true,
+    "ntasks": 1,
+    "cpus_per_task": 1,
+    "n_gpus": 1,
+    "gpu_type": "T4",
+    "device_id": "auto",
+    "mem": "16GB",
```
Comment (dennlinger) [inline exp_configs/cluster/openai-gpt-3.5-turbo.json:8]: Not sure if assigning GPUs is necessary here, but not really familiar with SLURM overall.

File: llm_inference.py [hunk 2]
```diff
@@ -420,6 +429,109 @@ def reshape_model_outputs(outputs: List[str], input_batch_size: int) -> List[Lis
...
+    """
+
+    def __init__(self, args: InferenceArguments):
+
+        self.args = args
+
+        from langchain.llms import Cohere, OpenAI
+        from langchain.chat_models import ChatOpenAI
+        
+        from api_secrets import COHERE_API_KEY, OPENAI_API_KEY
```
Comment (dennlinger) [inline llm_inference.py:444]: Is there a particular reason for putting the inputs within the model init? Only advantage I could see is that some of the dependencies may not be necessary?

File: llm_inference.py [hunk 3]
```diff
@@ -420,6 +429,109 @@ def reshape_model_outputs(outputs: List[str], input_batch_size: int) -> List[Lis
...
+            human_template="{text}"
+            human_message_prompt = HumanMessagePromptTemplate.from_template(human_template)
+            chat_prompt = ChatPromptTemplate.from_messages([human_message_prompt])
+            chain = LLMChain(llm=self.model, prompt=chat_prompt)
+            with get_openai_callback() as cb:
+                outputs = chain.run(inputs[0])
+                logger.info(f'{cb}'.replace("\n", ", "))
+                self.estimated_cost += cb.total_cost
+        else:
+            with get_openai_callback() as cb:
```
Comment (dennlinger) [inline llm_inference.py:523]: This is bound to throw an error if you call an Cohere model, probably?

**Review by dennlinger — COMMENTED** [review]
Looks good to me! There is only one minor actual concern regarding the openai callback handling with Cohere models, otherwise these are only design comments :)

Oh, and you also want to probably change the respective examples in the `README.md`, which (I believe) references the now no longer existing script.

File: llm_inference.py (hunk 2 above)
Comment (tannonk) [reply llm_inference.py:444]: Yeah, that was the only reason. I figured it 'encapuslated' the API model class a bit better.

File: exp_configs/cluster/openai-gpt-3.5-turbo.json (hunk 1 above)
Comment (tannonk) [reply exp_configs/cluster/openai-gpt-3.5-turbo.json:8]: This is only necessary for the evaluation part of the pipeline. A few of the metrics require a GPU, so it's specified here for ease of execution. The logic in the `run.py` doesn't apply this to the inference part for API models.

File: llm_inference.py (hunk 3 above)
Comment (tannonk) [reply llm_inference.py:523]: Nice spotting! You're right. This functionality is only available with OpenAI models atm (https://python.langchain.com/en/latest/modules/models/llms/examples/token_usage_tracking.html#how-to-track-token-usage). It's now fixed and tested with both OpenAI and Cohere models.

**Pull request merged**